mono/profiler/proflog.c

```diff
@@ -2490,6 +2541,517 @@ counters_and_perfcounters_sample (MonoProfiler *prof)
+	class_name = mono_type_get_name (mono_class_get_type (klass));
+
+	assembly_name = assembly_name ? assembly_name : "";
+	number_of_methods = mono_class_num_methods (klass);
+	fully_covered = class_methods->len;
+	/* We don't handle partial covered yet */
+	partially_covered = 0;
+
+	size += strlen (assembly_name) + 1;
+	size += strlen (class_name) + 1;
+	size += 30; /* number_of_methods, fully_covered, partially_covered */
+
+	logbuffer = ensure_logbuf (size);
+
+	ENTER_LOG (logbuffer, "coverage-class");
+	emit_byte (logbuffer, TYPE_COVERAGE_CLASS | TYPE_COVERAGE);
+	emit_string (logbuffer, assembly_name, strlen (assembly_name) + 1);
+	emit_string (logbuffer, class_name, strlen (class_name) + 1);
+	emit_uvalue (logbuffer, number_of_methods);
+	emit_uvalue (logbuffer, fully_covered);
+	emit_uvalue (logbuffer, partially_covered);
+	EXIT_LOG (logbuffer);
+
+	safe_send (prof, ensure_logbuf (0));
+
+	g_free (class_name);
+}
+
+static void
+get_coverage_for_image (MonoImage *image, int *number_of_methods, guint *fully_covered, int *partially_covered)
+{
+	GPtrArray *image_methods = g_hash_table_lookup (image_to_methods, image);
+
+	*number_of_methods = mono_image_get_table_rows (image, MONO_TABLE_METHOD);
+	if (image_methods) {
+		*fully_covered = image_methods->len;
+	} else {
+		*fully_covered = 0;
+	}
+
+	// FIXME: We don't handle partially covered yet.
+	*partially_covered = 0;
+}
+
+static void
+build_assembly_buffer (gpointer key, gpointer value, gpointer userdata)
+{
+	MonoAssembly *assembly = (MonoAssembly *)value;
+	MonoProfiler *prof = (MonoProfiler *)userdata;
+	MonoImage *image = mono_assembly_get_image (assembly);
+	LogBuffer *logbuffer;
+	const char *name, *guid, *filename;
+	int size = 1;
+	int number_of_methods = 0, partially_covered = 0;
+	guint fully_covered = 0;
+
+	name = mono_image_get_name (image);
+	guid = mono_image_get_guid (image);
+	filename = mono_image_get_filename (image);
+
+	name = name ? name : "";
+	guid = guid ? guid : "";
+	filename = filename ? filename : "";
+
+	get_coverage_for_image (image, &number_of_methods, &fully_covered, &partially_covered);
+
+	size += strlen (name) + 1;
+	size += strlen (guid) + 1;
+	size += strlen (filename) + 1;
+	size += 30; /* number_of_methods, fully_covered, partially_covered */
+	logbuffer = ensure_logbuf (size);
+
+	ENTER_LOG (logbuffer, "coverage-assemblies");
+	emit_byte (logbuffer, TYPE_COVERAGE_ASSEMBLY | TYPE_COVERAGE);
+	emit_string (logbuffer, name, strlen (name) + 1);
+	emit_string (logbuffer, guid, strlen (guid) + 1);
+	emit_string (logbuffer, filename, strlen (filename) + 1);
+	emit_uvalue (logbuffer, number_of_methods);
+	emit_uvalue (logbuffer, fully_covered);
+	emit_uvalue (logbuffer, partially_covered);
+	EXIT_LOG (logbuffer);
+
+	safe_send (prof, ensure_logbuf (0));
+}
+
+static void
+dump_coverage (MonoProfiler *prof)
+{
+	if (!coverage_initialized)
+		return;
+
+	COVERAGE_DEBUG(fprintf (stderr, "Coverage: Started dump\n");)
+	mono_mutex_lock (&coverage_mutex);
```

Comment:
I don't think we need a mutex lock here; afaik it's guaranteed that all threads will stop firing profiler events by the time we reach `log_shutdown`